# Introduction
This is an assignment for data cleaning and a little ETL for the Titanic data set.  The data  set used in this assignment was provided by the instructor. The requirements of the assignmet required the data set to be cleaned and 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
titanic_df = pd.read_excel('Week_04_HW4_Titanic_Data.xls')
print(titanic_df.head(2))

  Survived Passenger Class                            Name     Sex      Age  \
0      Yes           First   Allen, Miss. Elisabeth Walton  Female  29.0000   
1      Yes           First  Allison, Master. Hudson Trevor    Male   0.9167   

   No of Siblings or Spouses on Board  No of Parents or Children on Board  \
0                                   0                                   0   
1                                   1                                   2   

  Ticket Number  Passenger Fare    Cabin Port of Embarkation Life Boat  
0         24160        211.3375       B5         Southampton         2  
1        113781        151.5500  C22 C26         Southampton        11  


# Show some descriptive info about the data 
This code displays some data about the file the column names and the data types of the data that are stored in the columns.  The length of the dataframe will also be displayed.

In [3]:
titanic_df.info()   # info about the contents of the dataframe
len(titanic_df)     # Show the length of the dataframe  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 12 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Survived                            1309 non-null   object 
 1   Passenger Class                     1309 non-null   object 
 2   Name                                1309 non-null   object 
 3   Sex                                 1309 non-null   object 
 4   Age                                 1046 non-null   float64
 5   No of Siblings or Spouses on Board  1309 non-null   int64  
 6   No of Parents or Children on Board  1309 non-null   int64  
 7   Ticket Number                       1309 non-null   object 
 8   Passenger Fare                      1308 non-null   float64
 9   Cabin                               295 non-null    object 
 10  Port of Embarkation                 1307 non-null   object 
 11  Life Boat                           486 non

1309

# Does the file contain empty values?
We look to see which of the columns do not have any data associated with the column or feature of the data set.

In [4]:
#nan_titanic_sr = titanic_df.isna().sum()
titanic_df.isna().sum()

Survived                                 0
Passenger Class                          0
Name                                     0
Sex                                      0
Age                                    263
No of Siblings or Spouses on Board       0
No of Parents or Children on Board       0
Ticket Number                            0
Passenger Fare                           1
Cabin                                 1014
Port of Embarkation                      2
Life Boat                              823
dtype: int64

This files has no values for Age, Cabin, and Lifeboat.  These values will need to be looked into.  Some can be left alone but  the age may be changed to help with applications and predictions of future events.
# Copy of the orginal dataframe
We will create a copy of the orginal dataframe so we can process and clean this set of data.  We will drop the rows in the Age column that do not have any values. We can drop the other rows that do not not contain the data but this is not needed at this point.

In [5]:
#Copy the orginal Titanic dataframe
copy_titanic_df = titanic_df.copy(deep=True)
copy_titanic_df.head(2) # print the copy to see if it exists
len(copy_titanic_df)

1309

We will show the number of items in the dataframe both before and after the dropping of the Age rows that contain no informantion(NaN).  

In [6]:
copy_titanic_df.dropna(subset=['Age'], inplace=True)
print("Length after drop: ",len(copy_titanic_df))
print("After drop:", copy_titanic_df.isna().sum())

Length after drop:  1046
After drop: Survived                                0
Passenger Class                         0
Name                                    0
Sex                                     0
Age                                     0
No of Siblings or Spouses on Board      0
No of Parents or Children on Board      0
Ticket Number                           0
Passenger Fare                          1
Cabin                                 774
Port of Embarkation                     2
Life Boat                             629
dtype: int64


After the drop there are 1046 which is to be expected after dropping the 263 rows from the dataframe.
After this drop we can do calculations based on the data that would require the Age data. We can handle the data rows that did not have this data by doing additional research and adding this data to the rows.
# Split the dataframe 
We can split the dataframe into different files based on class so that we can look at each group separately, if we choose to do so.  The code below 

In [7]:
#Create the first class dataframe
first_class_df = copy_titanic_df[copy_titanic_df["Passenger Class"] == 'First']
print("First \n", first_class_df.head(2))
print("Total First Class: ",len(first_class_df))

#Create the second class dataframe
second_class_df = copy_titanic_df[copy_titanic_df["Passenger Class"] == 'Second']
print("\nSecond \n",second_class_df.head(2)) 
print("Total Second Class: ",len(second_class_df))

#Create third class dataframe
third_class_df = copy_titanic_df[copy_titanic_df["Passenger Class"] == 'Third']

print(third_class_df.head(2)) 
print("Total Third Class: ",len(third_class_df))


First 
   Survived Passenger Class                            Name     Sex      Age  \
0      Yes           First   Allen, Miss. Elisabeth Walton  Female  29.0000   
1      Yes           First  Allison, Master. Hudson Trevor    Male   0.9167   

   No of Siblings or Spouses on Board  No of Parents or Children on Board  \
0                                   0                                   0   
1                                   1                                   2   

  Ticket Number  Passenger Fare    Cabin Port of Embarkation Life Boat  
0         24160        211.3375       B5         Southampton         2  
1        113781        151.5500  C22 C26         Southampton        11  
Total First Class:  284

Second 
     Survived Passenger Class                                   Name     Sex  \
323       No          Second                    Abelson, Mr. Samuel    Male   
324      Yes          Second  Abelson, Mrs. Samuel (Hannah Wizosky)  Female   

      Age  No of Siblings or Sp

To make sure that all the passengers are have been inserted into the appropriate data frame we can look at the lengths of the dataframe.

In [8]:
#Output the length of the dataframes and the  total of all dataframes
print("Total from COPY: ", len(copy_titanic_df))
print("Total from First: ", len(first_class_df))
print("Total from SECOND: ", len(second_class_df))
print("Total from THIRD: ", len(third_class_df))
total = len(first_class_df) +  len(second_class_df) +len(third_class_df)
print("Total is : ", total)

Total from COPY:  1046
Total from First:  284
Total from SECOND:  261
Total from THIRD:  501
Total is :  1046


# Export for future use.
Export the dataframes based on passenger class to a Excel file for later use. 


In [9]:
# Export the partitioned data to Excel file.
#first_class_df.to_excel(r'first_class.xlsx', index=False)
#second_class_df.to_excel(r'second_class.xlsx', index=False)
#third_class_df.to_excel(r'third_class.xlsx', index=False)